### PR TITLE
have addon_factory produce complete addons that have metadata to be public

### PR DESCRIPTION
**Pull request by eviljeff**
This is part of the work to stop setting addon.status directly.

### CODE REVIEW

**Comment by eviljeff** [comment]
addon_factory might have to be re-thought as part of that work anyway - generating an addon with status=STATUS_NULL won't be a thing exactly, rather an add-on will be STATUS_NULL because it has no non-disable listed versions, so we'll need to generate as appropriate.

File: src/olympia/amo/tests/__init__.py [hunk 1]
```diff
@@ -766,6 +794,8 @@ def version_factory(file_kw=None, **kw):
     max_app_version = kw.pop('max_app_version', '5.0.99')
     version = kw.pop('version', '%.1f' % random.uniform(0, 2))
     application = kw.pop('application', amo.FIREFOX.id)
+    if not kw.get('license') and not kw.get('license_id'):
+        kw['license'] = license_factory(**(kw.get('license_kw', {})))
```
Comment (diox) [inline src/olympia/amo/tests/__init__.py]: Should we create a default re-usable license instead, to avoid stuffing the test db with useless licenses when creating multiple add-ons ?

File: src/olympia/amo/tests/__init__.py [hunk 2]
```diff
@@ -717,11 +729,27 @@ def collection_factory(**kw):
...
+    return License.objects.create(**data)
+
+
 def file_factory(**kw):
     v = kw['version']
+    filename = kw.pop('filename', '%s-%s' % (v.addon_id, v.id))
     status = kw.pop('status', amo.STATUS_PUBLIC)
     platform = kw.pop('platform', amo.PLATFORM_ALL.id)
-    f = File.objects.create(filename='%s-%s' % (v.addon_id, v.id),
+    f = File.objects.create(filename=filename,
```
Comment (diox) [inline src/olympia/amo/tests/__init__.py]: Please take the opportunity to rename the `f` and `v` variable :crying_cat_face:

File: src/olympia/amo/tests/__init__.py [hunk 3]
```diff
@@ -678,6 +682,14 @@ def addon_factory(
     for user in users:
         addon.addonuser_set.create(user=user)
 
+    application = version_kw.get('application', amo.FIREFOX.id)
+    if not category:
+        category = Category.from_static_category(
+            next(CATEGORIES[application][type_].itervalues()))
```
Comment (diox) [inline src/olympia/amo/tests/__init__.py]: Mmmm, this does not ensure that the category does not exist yet, maybe it should (through `.get_or_create` with the right `defaults` instead of `from_static_category` maybe ?)

Comment (diox) [inline src/olympia/amo/tests/__init__.py]: Also, how about just picking a category at random ? Not sure if it's a good idea...

File: src/olympia/amo/tests/__init__.py (hunk 1 above)
Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: Re-using the same license for different add-ons might create some weird edge cases. Could re-use from an earlier version in the same add-on though (like `Version.from_upload` does)

Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: though that wouldn't really address your concern as I don't think we create multiple versions on the same add-on that often.

File: src/olympia/amo/tests/__init__.py (hunk 3 above)
Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: Don't use `from_static_category`?  Static categories are the norm in prod AMO I thought.

File: src/olympia/amo/tests/__init__.py (hunk 1 above)
Comment (diox) [inline src/olympia/amo/tests/__init__.py]: Well, it shouldn't create more edge cases since we do have the concept of built-in licenses ? But if it's too complex let's drop it

File: src/olympia/amo/tests/__init__.py (hunk 3 above)
Comment (diox) [inline src/olympia/amo/tests/__init__.py]: They are, but `from_static_category()` only really does `return cls(**static_category.__dict__)` - It's used by some internal code to return fake `Category` instances that are never meant to be saved. 

Some tests use that as a shortcut, but you can easily do it yourself with something like (untested) `get_or_create(Category, id=data['id'], defaults=data)` where `data` is one of the items in `CATEGORIES[application][type_]`

File: src/olympia/amo/tests/__init__.py (hunk 1 above)
Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: could create it as a built-in license, sure.

File: src/olympia/amo/tests/__init__.py (hunk 3 above)
Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: https://github.com/mozilla/addons-server/pull/4709/commits/b7a55bbb4fe67a2e8a3e2dadf5861595a9f3d114

File: src/olympia/amo/tests/__init__.py (hunk 1 above)
Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: https://github.com/mozilla/addons-server/pull/4709/commits/b7a55bbb4fe67a2e8a3e2dadf5861595a9f3d114

File: src/olympia/amo/tests/__init__.py [hunk 4]
```diff
@@ -764,22 +792,31 @@ def version_factory(file_kw=None, **kw):
...
-    version = kw.pop('version', '%.1f' % random.uniform(0, 2))
+    version_str = kw.pop('version', '%.1f' % random.uniform(0, 2))
     application = kw.pop('application', amo.FIREFOX.id)
-    v = Version.objects.create(version=version, **kw)
-    v.created = v.last_updated = _get_created(kw.pop('created', 'now'))
-    v.save()
+    if not kw.get('license') and not kw.get('license_id'):
+        # Is there a built-in one we can use?
+        builtins = License.objects.builtins()
+        if builtins.count():
```
Comment (diox) [inline src/olympia/amo/tests/__init__.py]: `builtins.exists()`

Comment (eviljeff) [inline src/olympia/amo/tests/__init__.py]: https://github.com/mozilla/addons-server/pull/4709/commits/c695b8bf92ebd82455d03b8c023974bcc88ae0af
